solution/ui/prototype/src/utilities/api.js

```diff
@@ -0,0 +1,371 @@
+            } catch (err) {
+                if (httpStatus >= 400) {
+                    if (
+                        httpStatus >= 501 &&
+                        retryCount < config.maxRetryCount
+                    ) {
+                        let backoff = retryCount * retryCount;
+                        if (backoff < 1) backoff = 1;
+
+                        return Promise.resolve()
+                            .then(() =>
+                                console.log(
+                                    `Retrying count = ${retryCount}, Backoff = ${backoff}`
+                                )
+                            )
+                            .then(() => delay(backoff))
+                            .then(() =>
+                                fetchJson(url, options, retryCount + 1)
+                            );
+                    }
+                    throw parseError({
+                        message: text,
+                        status: httpStatus,
+                    });
+                } else {
+                    throw parseError(
+                        new Error("The server did not return a json response.")
+                    );
+                }
+            }
+
+            return json;
+        })
+        .then((json) => {
+            if (_isBoolean(isOk) && !isOk) {
+                throw parseError({ ...json, status: httpStatus });
+            } else {
+                return json;
+            }
+        });
+}
+
+// ---------- helper functions ---------------
+
+function httpApiMock(verb, urlPath, { data, params, response } = {}) {
+    console.log(`Performing an HTTP ${verb} to ${config.apiPath}/${urlPath}`);
+    data && console.log("DATA: ", data);
+    params && console.log("PARAMS: ", params);
+    response && console.log("RESPONSE: ", response);
+    return response;
+}
+
+function httpApiGet(urlPath, { params } = {}) {
+    return fetchJson(`${config.apiPath}/${urlPath}`, {
+        method: "GET",
+        headers: authHeader(token),
+        params,
+    });
+}
+
+function httpApiPost(urlPath, { data = {}, params } = {}) {
+    console.log(data);
+    return fetchJson(`${config.apiPath}/${urlPath}`, {
+        method: "POST",
+        headers: authHeader(token),
+        params,
+        body: JSON.stringify(data),
+    });
+}
+
+// eslint-disable-next-line no-unused-vars
+function httpApiPut(urlPath, { data, params } = {}) {
+    return fetchJson(`${config.apiPath}/${urlPath}`, {
+        method: "PUT",
+        headers: authHeader(token),
+        params,
+        body: JSON.stringify(data),
+    });
+}
+
+// eslint-disable-next-line no-unused-vars
+function httpApiDelete(urlPath, { data, params } = {}) {
+    return fetchJson(`${config.apiPath}/${urlPath}`, {
+        method: "DELETE",
+        headers: authHeader(token),
+        params,
+        body: JSON.stringify(data),
+    });
+}
+
+// ---------- api calls ---------------
+
+const login = (username, password) => {
```

Comment:
This and the next method can probably get deleted.

Reply:
deleted